driver/transform-dsl.md

```diff
@@ -0,0 +1,549 @@
+S  := A.Check(T)
+T' := B.Construct(S)
+
+// reverse transform
+S' := B.Check(T')
+S == S'
+T == A.Construct(S')
+```
+
+## List of operations
+
+### Values
+
+#### Any value
+
+In case any value can match the transformation, it should be stored into
+a named variable: `Var("x")`. Any name can be used for the variable as
+long as it is unique in this transformation.
+
+When writing mappings, variable with the same name should exist in the
+second transformation.
+
+#### Equal
+
+To check value for exact match, `Is(v)` operation is used.
+
+SDK defines helpers for common types: `String(v)`, `Int(v)`, etc.
+
+Null values can be checked with `Is(nil)`.
+
+#### Equal or nil
+
+To check any node for nil or specific rule `Opt` is used:
+
+```
+Opt("is_nil", Is(42))
+```
+
+`Opt` needs a named variable to store it's state ("is_nil" in this case).
+
+#### Value in
+
+To check that value equal any value in the list, two operations should be
+executed:
+
+1) Check that value is in the list.
+
+2) Store the value into a variable.
+
+To express this, `Check` helper can be used:
+
+```golang
+Check(
+    // condition
+    In(
+        uast.String("A"),
+        uast.String("B"),
+    ),
+
+    // action
+    Var("x"),
+)
+```
+
+#### Not equal, not in, etc
+
+To check value is not equal `Check` can be combined with `Not`:
+
+```golang
+Check(
+    // condition
+    Not(String("A")),
+
+    // store the value
+    Var("x"),
+)
+```
+
+Note that value can still be of any type and type constraint of `String`
+will not affect the variable.
+
+The same way a value can be checked for "not in" constraint:
+
+```golang
+Check(
+    Not(In(
+            uast.String("A"),
+            uast.String("B"),
+    )),
+    // store the value
+    Var("x"),
+)
+```
+
+### Objects
+
+#### Empty object
+
+To represent an empty object, `Obj{}` can be used.
+
+#### Object or nil
+
+To check any node for nil or specific rule `Opt` is used:
+
+```
+Opt("is_nil", Obj{ ... })
+```
+
+`Opt` needs a named variable to store it's state ("is_nil" in this case).
+
+#### All fields must exist
+
+For object that always have a specific set of fields, `Obj` operation
+can be used:
+
+```golang
+Obj{
+    "type": String("Ident"),
+    "name": Var("name"),
+}
+```
+
+Since this operation is represented by a map, order of operations is not
+preserved. Operations will be executed in order of field name sorting.
+This might affect operations that requres a variable to be defined, such
+as `Lookup` and `If`. To control execution order, use `Fields`.
+
+#### Specific field options
+
+To get more control of how object fields are checked, use `Fields`:
+
+```golang
+Fields{
+    {Name: "type", Op: String("If")},
+    {Name: "then": Op: Var("then")},
+    {Name: "else": Op: Var("then"), Optional: "else_exists"},
+}
+```
+
+For each optional field a variable should be provided to store the state.
+Optional field means that they may not exist. For checking nil fields see
+`Opt` operation.
+
```

Comment:
Since Field is pretty important, could use one of those expanded examples with the json representations of the objects before and after the transformation.